vnext/Microsoft.ReactNative.Cxx/JSI/JsiAbiApi.h

```diff
@@ -0,0 +1,318 @@
+};
+
+// A wrapper for ABI-safe JsiPreparedJavaScript.
+struct JsiPreparedJavaScriptWrapper : facebook::jsi::PreparedJavaScript {
+  JsiPreparedJavaScriptWrapper(JsiPreparedJavaScript const &preparedScript) noexcept;
+  ~JsiPreparedJavaScriptWrapper() noexcept;
+  JsiPreparedJavaScript const &Get() const noexcept;
+
+ private:
+  JsiPreparedJavaScript m_preparedScript;
+};
+
+// An ABI-safe wrapper for facebook::jsi::HostObject.
+struct JsiHostObjectWrapper : implements<JsiHostObjectWrapper, IJsiHostObject> {
+  JsiHostObjectWrapper(std::shared_ptr<facebook::jsi::HostObject> &&hostObject) noexcept;
+  ~JsiHostObjectWrapper() noexcept;
+
+  JsiValueRef GetProperty(JsiRuntime const &runtime, JsiPropertyIdRef const &name);
+  void SetProperty(JsiRuntime const &runtime, JsiPropertyIdRef const &name, JsiValueRef const &value);
+  Windows::Foundation::Collections::IVector<JsiPropertyIdRef> GetPropertyIds(JsiRuntime const &runtime);
+
+  static void RegisterHostObject(JsiObjectRef const &objectData, JsiHostObjectWrapper *hostObject) noexcept;
+  static bool IsHostObject(JsiObjectRef const &objectData) noexcept;
+  static std::shared_ptr<facebook::jsi::HostObject> GetHostObject(JsiObjectRef const &objectData) noexcept;
+
+ private:
+  std::shared_ptr<facebook::jsi::HostObject> m_hostObject;
+  JsiObjectRef m_objectData{};
+
+  static std::mutex s_mutex;
+  static std::map<uint64_t, JsiHostObjectWrapper *> s_objectDataToObjectWrapper;
+};
+
+// The function object that wraps up the facebook::jsi::HostFunctionType
+struct JsiHostFunctionWrapper {
+  // We only support new and move constructors.
+  JsiHostFunctionWrapper(facebook::jsi::HostFunctionType &&hostFunction, uint32_t functionId) noexcept;
+  JsiHostFunctionWrapper(JsiHostFunctionWrapper &&other) noexcept;
+  ~JsiHostFunctionWrapper() noexcept;
+
+  // Disable other ways to construct or modify the wrapper.
+  JsiHostFunctionWrapper &operator=(JsiHostFunctionWrapper &&other) = delete;
+  JsiHostFunctionWrapper(JsiHostFunctionWrapper const &other) = delete;
+  JsiHostFunctionWrapper &operator=(JsiHostFunctionWrapper const &other) = delete;
+
+  JsiValueRef operator()(JsiRuntime const &runtime, JsiValueRef const &thisArg, array_view<JsiValueRef const> args);
+
+  static uint32_t GetNextFunctionId() noexcept;
+  static void RegisterHostFunction(uint32_t functionId, JsiObjectRef const &func) noexcept;
+  static bool IsHostFunction(JsiObjectRef const &func) noexcept;
+  static facebook::jsi::HostFunctionType &GetHostFunction(JsiObjectRef const &func) noexcept;
+
+ private:
+  facebook::jsi::HostFunctionType m_hostFunction;
+  JsiObjectRef m_functionData{};
+  uint32_t m_functionId{};
+
+  static std::mutex s_functionMutex;
+  static std::atomic<uint32_t> s_functionIdGenerator;
+  static std::map<uint32_t, JsiHostFunctionWrapper *> s_functionIdToFunctionWrapper;
+  static std::map<uint64_t, JsiHostFunctionWrapper *> s_functionDataToFunctionWrapper;
```

Comment:
Minor point: It looks like these could be `unordered_map` along with some of the other static maps in this PR.

Reply:
Thanks! I would like to remove it. It is a work-around for the issue with the Chakra JSI implementation (getHostFunction() is not implemented and throws an exception). I am fixing the Chakra JSI in #5710 and after that I should be able to use the HostObject and HostFunction directly without needing the mappings.